erroneously error: Block-scoped variable used before its declaration

**TypeScript Version:**  2.0.5 

**Code**

```ts
// A.ts
namespace ts {
    export function printVersion():void {
        console.log("Version: " + sys.version);
    }
}

// B.ts
namespace ts {
    export let sys:{version:string} = {version: "2.0.5"};
    ts.printVersion();
}

// tsconfig.json
{
  "compilerOptions": {
    "target": "es5",
    "outFile" : "test.js"
  },
  "files": [
    "A.ts",
    "B.ts"
  ]
}
```
When I try to compile these two simple files (A.ts and B.ts) into one test.js file. The compiler reports an error: 

```
A.ts(4,35): error TS2448: Block-scoped variable 'sys' used before its declaration.
```
But if I merge these two files into one like this:

```
// Single.ts
namespace ts {
    export function printVersion():void {
        console.log("Version: " + sys.version);
    }
    export let sys:{version:string} = {version: "2.0.5"};
    ts.printVersion();
}
```
The compiler does not report this error.

Obviously the usage of `sys.version` is deferred, because it is inside a function. So it should be legal even if the usage and declaration are in separate files. Then I checked the source code of TypeScript, I think the code that caused this problem might be here:

src/compiler/checker.ts line:585
```
 function isBlockScopedNameDeclaredBeforeUse(declaration: Declaration, usage: Node): boolean {
            const declarationFile = getSourceFileOfNode(declaration);
            const useFile = getSourceFileOfNode(usage);
            if (declarationFile !== useFile) {
                if ((modulekind && (declarationFile.externalModuleIndicator || useFile.externalModuleIndicator)) ||
                    (!compilerOptions.outFile && !compilerOptions.out)) {
                    // nodes are in different files and order cannot be determines
                    return true;
                }

                const sourceFiles = host.getSourceFiles();
                return indexOf(sourceFiles, declarationFile) <= indexOf(sourceFiles, useFile);
            }

            if (declaration.pos <= usage.pos) {
                // declaration is before usage
                // still might be illegal if usage is in the initializer of the variable declaration
                return declaration.kind !== SyntaxKind.VariableDeclaration ||
                    !isImmediatelyUsedInInitializerOfBlockScopedVariable(<VariableDeclaration>declaration, usage);
            }

            // declaration is after usage
            // can be legal if usage is deferred (i.e. inside function or in initializer of instance property)
            return isUsedInFunctionOrNonStaticProperty(declaration, usage);
}
```

In the `isBlockScopedNameDeclaredBeforeUse`  function , if the `declarationFile`  and `useFile` are not the same,  it forgets to check whether the usage is deferred.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -591,7 +591,11 @@ namespace ts {
                     // nodes are in different files and order cannot be determines
                     return true;
                 }
-
+                // declaration is after usage
+                // can be legal if usage is deferred (i.e. inside function or in initializer of instance property)
+                if(isUsedInFunctionOrNonStaticProperty(usage)){
+                    return true;
+                }
                 const sourceFiles = host.getSourceFiles();
                 return indexOf(sourceFiles, declarationFile) <= indexOf(sourceFiles, useFile);
             }
@@ -605,7 +609,8 @@ namespace ts {
 
             // declaration is after usage
             // can be legal if usage is deferred (i.e. inside function or in initializer of instance property)
-            return isUsedInFunctionOrNonStaticProperty(declaration, usage);
+            const container = getEnclosingBlockScopeContainer(declaration);
+            return isUsedInFunctionOrNonStaticProperty(usage, container);
 
             function isImmediatelyUsedInInitializerOfBlockScopedVariable(declaration: VariableDeclaration, usage: Node): boolean {
                 const container = getEnclosingBlockScopeContainer(declaration);
@@ -634,8 +639,7 @@ namespace ts {
                 return false;
             }
 
-            function isUsedInFunctionOrNonStaticProperty(declaration: Declaration, usage: Node): boolean {
-                const container = getEnclosingBlockScopeContainer(declaration);
+            function isUsedInFunctionOrNonStaticProperty(usage: Node, container? :Node): boolean {
                 let current = usage;
                 while (current) {
                     if (current === container) {

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -593,7 +593,7 @@ namespace ts {
                 }
                 // declaration is after usage
                 // can be legal if usage is deferred (i.e. inside function or in initializer of instance property)
-                if(isUsedInFunctionOrNonStaticProperty(usage)){
+                if (isUsedInFunctionOrNonStaticProperty(usage)) {
                     return true;
                 }
                 const sourceFiles = host.getSourceFiles();
@@ -639,7 +639,7 @@ namespace ts {
                 return false;
             }
 
-            function isUsedInFunctionOrNonStaticProperty(usage: Node, container? :Node): boolean {
+            function isUsedInFunctionOrNonStaticProperty(usage: Node, container?: Node): boolean {
                 let current = usage;
                 while (current) {
                     if (current === container) {

--- a/tests/baselines/reference/useBeforeDeclaration.js
+++ b/tests/baselines/reference/useBeforeDeclaration.js
@@ -1,0 +1,38 @@
+//// [tests/cases/compiler/useBeforeDeclaration.ts] ////
+
+//// [A.ts]
+
+namespace ts {
+    export function printVersion():void {
+        log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+    }
+
+    export function log(info:string):void {
+
+    }
+}
+
+//// [B.ts]
+namespace ts {
+
+    export let sys:{version:string} = {version: "2.0.5"};
+
+}
+
+
+
+//// [test.js]
+var ts;
+(function (ts) {
+    function printVersion() {
+        log("Version: " + ts.sys.version); // the call of sys.version is deferred, should not report an error.
+    }
+    ts.printVersion = printVersion;
+    function log(info) {
+    }
+    ts.log = log;
+})(ts || (ts = {}));
+var ts;
+(function (ts) {
+    ts.sys = { version: "2.0.5" };
+})(ts || (ts = {}));

--- a/tests/baselines/reference/useBeforeDeclaration.symbols
+++ b/tests/baselines/reference/useBeforeDeclaration.symbols
@@ -1,0 +1,34 @@
+=== tests/cases/compiler/A.ts ===
+
+namespace ts {
+>ts : Symbol(ts, Decl(A.ts, 0, 0), Decl(B.ts, 0, 0))
+
+    export function printVersion():void {
+>printVersion : Symbol(printVersion, Decl(A.ts, 1, 14))
+
+        log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+>log : Symbol(log, Decl(A.ts, 4, 5))
+>sys.version : Symbol(version, Decl(B.ts, 2, 20))
+>sys : Symbol(sys, Decl(B.ts, 2, 14))
+>version : Symbol(version, Decl(B.ts, 2, 20))
+    }
+
+    export function log(info:string):void {
+>log : Symbol(log, Decl(A.ts, 4, 5))
+>info : Symbol(info, Decl(A.ts, 6, 24))
+
+    }
+}
+
+=== tests/cases/compiler/B.ts ===
+namespace ts {
+>ts : Symbol(ts, Decl(A.ts, 0, 0), Decl(B.ts, 0, 0))
+
+    export let sys:{version:string} = {version: "2.0.5"};
+>sys : Symbol(sys, Decl(B.ts, 2, 14))
+>version : Symbol(version, Decl(B.ts, 2, 20))
+>version : Symbol(version, Decl(B.ts, 2, 39))
+
+}
+
+

--- a/tests/baselines/reference/useBeforeDeclaration.types
+++ b/tests/baselines/reference/useBeforeDeclaration.types
@@ -1,0 +1,39 @@
+=== tests/cases/compiler/A.ts ===
+
+namespace ts {
+>ts : typeof ts
+
+    export function printVersion():void {
+>printVersion : () => void
+
+        log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+>log("Version: " + sys.version) : void
+>log : (info: string) => void
+>"Version: " + sys.version : string
+>"Version: " : "Version: "
+>sys.version : string
+>sys : { version: string; }
+>version : string
+    }
+
+    export function log(info:string):void {
+>log : (info: string) => void
+>info : string
+
+    }
+}
+
+=== tests/cases/compiler/B.ts ===
+namespace ts {
+>ts : typeof ts
+
+    export let sys:{version:string} = {version: "2.0.5"};
+>sys : { version: string; }
+>version : string
+>{version: "2.0.5"} : { version: string; }
+>version : string
+>"2.0.5" : "2.0.5"
+
+}
+
+

--- a/tests/cases/compiler/useBeforeDeclaration.ts
+++ b/tests/cases/compiler/useBeforeDeclaration.ts
@@ -1,20 +1,13 @@
-// @fileName: tsconfig.json
-// {
-//     "compilerOptions": {
-//     "target": "es5",
-//         "outFile" : "test.js"
-//     },
-//     "files": [
-//     "A.ts",
-//     "B.ts"
-//     ]
-// }
-
+// @outFile: test.js
 
 // @fileName: A.ts
 namespace ts {
     export function printVersion():void {
-        console.log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+        log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+    }
+
+    export function log(info:string):void {
+
     }
 }
 
@@ -23,6 +16,5 @@ namespace ts {
 
     export let sys:{version:string} = {version: "2.0.5"};
 
-    ts.printVersion();
 }
 

--- a/tests/cases/compiler/useBeforeDeclaration.ts
+++ b/tests/cases/compiler/useBeforeDeclaration.ts
@@ -1,0 +1,28 @@
+// @fileName: tsconfig.json
+// {
+//     "compilerOptions": {
+//     "target": "es5",
+//         "outFile" : "test.js"
+//     },
+//     "files": [
+//     "A.ts",
+//     "B.ts"
+//     ]
+// }
+
+
+// @fileName: A.ts
+namespace ts {
+    export function printVersion():void {
+        console.log("Version: " + sys.version);  // the call of sys.version is deferred, should not report an error.
+    }
+}
+
+// @fileName: B.ts
+namespace ts {
+
+    export let sys:{version:string} = {version: "2.0.5"};
+
+    ts.printVersion();
+}
+

